app/controllers/admin/badges_controller.rb

```diff
@@ -52,37 +54,79 @@ def mass_award
     end
 
     replace_badge_owners = params[:replace_badge_owners] == 'true'
-    BadgeGranter.revoke_all(badge) if replace_badge_owners
+    grant_existing_holders = params[:grant_existing_holders] == 'true'
+    if grant_existing_holders && !badge.multiple_grant?
+      render_json_error(
+        I18n.t('badges.mass_award.errors.cant_grant_multiple_times', badge_name: badge.display_name),
+        status: 422
+      )
+      return
+    end
 
-    batch_number = 1
     line_number = 1
-    batch = []
-
+    entries = []
     File.open(csv_file) do |csv|
-      mode = Email.is_valid?(CSV.parse_line(csv.first).first) ? 'email' : 'username'
-      csv.rewind
+      csv.each_line do |line|
```

Comment:
maybe use `.with_index` here, so we don't have all the `line_number` code ?

Reply:
We could do that, but we'd still need to keep `line_variable` variable because if we encounter an error while reading the CSV file, we report the problematic line number to the user (see line 131 in the controller).